### PR TITLE
ddl: make "add foreign key" step in multiple schema versions

### LINKED ISSUE #38678: add foreign key need multi-schema change
## Dev Task

At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create foreign key in one schema version, since this may break foreign key constraint, such as delete reference table
without foreign key constrain check in child table.

```sql
-- In TiDB-1 and Schema Version is 1
insert into t_has_foreign_key values (1, 1);
-- In TiDB-0 and  Schema Version is 0
delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
```

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #38678 

At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create foreign key in one schema version, since this may break foreign key constraint, such as delete reference table
without foreign key constrain check in child table.

```sql
-- In TiDB-1 and Schema Version is 1
insert into t_has_foreign_key values (1, 1);
-- In TiDB-0 and  Schema Version is 0
delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
```

So The DDL owner handle add foreign key constrain step should be:

1. None -> Write Only: add foreign key constrain which state is `write-only` into table.
3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
   ```sql
   select count(*) from t2 where t2.a not in (select id from t1);
   ```
   The expected result is `0`, otherwise, an error is returned and cancel the ddl job.
4. Write Reorg -> Public: update the foreign key constraint state to `public`.

A problem is, How the DML treat the foreign key with on delete/update cascade behaviour which state is non-public?
Here is an example:

```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int, index(a));
insert into t1 values (1,1);
insert into t2 values (1,1);
alter  table t2 add constraint fk_1 foreign key (a) references t1(id) ON DELETE CASCADE;
```

The schema change of foreign key `fk_1` is from `None` -> `Write-Only` -> `Write-Reorg` -> `Public`。
When the foreign key `fk_1` in `Write-Only` state, a DML request has come to be processed:

```sql
delete from t1 where id = 1;
```

Then, TiDB shouldn't do cascade delete for foreign key `fk_1` in state `Write-Only`, since the `Add Foreign Key` DDL job maybe
failed in `Write-Reorg` state and rollback the DDL job. But it is hard to rollback the cascade deleted executed before.

So, when execute DML with `non-publick` foreign key, TiDB will do foreign key constraint check instead of foreign key cascade behaviour.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/foreign_key.go [hunk 1]
```diff
@@ -44,29 +44,46 @@ func (w *worker) onCreateForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job) (ve
...
+	switch job.SchemaState {
 	case model.StateNone:
-		// We just support record the foreign key, so we just make it public.
-		// none -> public
-		fkInfo.State = model.StatePublic
-		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != fkInfo.State)
+		err = checkAddForeignKeyValidInOwner(d, t, job, job.SchemaName, tblInfo, &fkInfo, fkCheck)
+		if err != nil {
+			return ver, err
+		}
```
Comment (tangenta) [inline ddl/foreign_key.go:52]: Should the job be cancelled?

Comment (crazycs520) [reply ddl/foreign_key.go:52]: done.

**Review by tangenta — APPROVED** [review]

**Review by lcwangchao — APPROVED** [review]

**Pull request merged**